src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs

```diff
@@ -57,50 +57,56 @@ LogicalMessage[] Extract(IncomingMessage physicalMessage)
             if (IsControlMessage(physicalMessage))
             {
                 log.Debug("Received a control message. Skipping deserialization as control message data is contained in the header.");
-                return NoMessagesFound;
+                return Array.Empty<LogicalMessage>();
             }
 
             if (physicalMessage.Body.Length == 0)
             {
                 log.Debug("Received a message without body. Skipping deserialization.");
-                return NoMessagesFound;
+                return Array.Empty<LogicalMessage>();
             }
 
-            var messageMetadata = new List<MessageMetadata>();
-
+            Type[] messageTypes = Array.Empty<Type>();
             if (physicalMessage.Headers.TryGetValue(Headers.EnclosedMessageTypes, out var messageTypeIdentifier))
             {
-                foreach (var messageTypeString in messageTypeIdentifier.Split(EnclosedMessageTypeSeparator))
-                {
-                    var typeString = messageTypeString;
-
-                    if (DoesTypeHaveImplAddedByVersion3(typeString))
-                    {
-                        continue;
-                    }
-
-                    var metadata = messageMetadataRegistry.GetMessageMetadata(typeString);
-
-                    if (metadata == null)
+                messageTypes = enclosedMessageTypesStringToMessageTypes.GetOrAdd(messageTypeIdentifier,
```

Comment:
Super minor but `messageTypeIndentifier` might be better as `enclosedMessageTypesValue`

Reply:
Done